### PR TITLE
Avalonia - Couple fixes and improvements to vulkan

### PR DESCRIPTION
**Pull request by emmauss**
1. Fix up crash that occurs when you toggle fullscreen
2. Add fallback to OpenGl if vulkan is not available.
3. Add swapchain present mode control to gtk
4. Fix screenshot on avalonia vulkan 
5. Fix favorite not being saved on avalonia

### CODE REVIEW

**Comment by TSRBerry** [comment]
Both the settings and the main window have this sort of transparent effect at the bottom and right side.
image

Clicking the left side of the controller type dropdown will actually click the reset button for the input device.
image

Running a game in fullscreen looks like this:
image

File: Ryujinx.Ava/Program.cs [hunk 1]
```diff
@@ -181,6 +180,18 @@ private static void Initialize(string[] args)
...
 
+            if (UseVulkan)
+            {
+                if (VulkanRenderer.GetPhysicalDevices().Length == 0)
+                {
+                    UseVulkan = false;
+
+                    ConfigurationState.Instance.Graphics.GraphicsBackend.Value = GraphicsBackend.OpenGl;
+
+                    Logger.Warning?.PrintMsg(LogClass.Application, $"A suitable vulkan physical device is not available. Falling back to OpenGl");
```
Comment (gdkchan) [inline Ryujinx.Ava/Program.cs:191]:
```suggestion
                    Logger.Warning?.PrintMsg(LogClass.Application, $"A suitable Vulkan physical device is not available. Falling back to OpenGL");
```

File: Ryujinx.Ava/Ui/Backend/Vulkan/VulkanCommandBufferPool.cs [hunk 2]
```diff
@@ -101,6 +110,20 @@ internal unsafe VulkanCommandBuffer(VulkanDevice device, VulkanCommandBufferPool
                 device.Api.CreateFence(device.InternalHandle, fenceCreateInfo, null, out _fence);
             }
 
+            public void WaitForFence()
+            {
+                if (!_isDisposed)
```
Comment (gdkchan) [inline Ryujinx.Ava/Ui/Backend/Vulkan/VulkanCommandBufferPool.cs:115]: You could change it to:
```cs
if (_isDisposed)
{
    return;
}
```
To avoid so much nesting.

File: Ryujinx.Ava/Ui/Backend/Vulkan/VulkanCommandBufferPool.cs [hunk 3]
```diff
@@ -171,11 +194,19 @@ public void Submit()
                 _commandBufferPool.DisposeCommandBuffer(this);
             }
 
-            public void Dispose()
+            public unsafe void Dispose()
             {
-                _device.Api.WaitForFences(_device.InternalHandle, 1, _fence, true, ulong.MaxValue);
-                _device.Api.FreeCommandBuffers(_device.InternalHandle, _commandBufferPool._commandPool, 1, InternalHandle);
-                _device.Api.DestroyFence(_device.InternalHandle, _fence, Span<AllocationCallbacks>.Empty);
```
Comment (gdkchan) [inline Ryujinx.Ava/Ui/Backend/Vulkan/VulkanCommandBufferPool.cs:178]: Why was this changed? Passing the empty span allows removing the `unsafe` keyword from the function.

File: Ryujinx.Ava/Ui/Backend/Vulkan/VulkanImage.cs [hunk 4]
```diff
@@ -148,20 +148,18 @@ internal void TransitionLayout(ImageLayout destinationLayout, AccessFlags destin
...
-        }
-
         public unsafe void Dispose()
         {
-            _device.Api.DestroyImageView(_device.InternalHandle, _imageView.Value, null);
-            _device.Api.DestroyImage(_device.InternalHandle, InternalHandle.Value, null);
-            _device.Api.FreeMemory(_device.InternalHandle, _imageMemory, null);
+            if (InternalHandle != null)
+            {
+                _device.Api.DestroyImageView(_device.InternalHandle, _imageView.Value, null);
```
Comment (gdkchan) [inline Ryujinx.Ava/Ui/Backend/Vulkan/VulkanImage.cs:155]: Could probably be made "safe" by passing empty span rather than null as the allocator argument.

File: Ryujinx.Ava/Ui/Controls/VulkanRendererControl.cs [hunk 5]
```diff
@@ -51,19 +86,42 @@ internal override void MakeCurrent(SwappableNativeWindowBase window)
 
         internal override void Present(object image)
         {
-            Dispatcher.UIThread.InvokeAsync(() =>
+            Image = image;
+
+            lock (_imagesInFlight)
```
Comment (gdkchan) [inline Ryujinx.Ava/Ui/Controls/VulkanRendererControl.cs:91]: If you need to lock it anyway, why are you using a `ConcurrentQueue`? The idea of the `ConcurrentQueue` is that you don't need to use locks since its already thread safe. If you need to lock might aswell just use a regular `Queue`.

File: Ryujinx.Ava/Ui/ViewModels/MainWindowViewModel.cs [hunk 6]
```diff
@@ -929,6 +930,13 @@ public static void OpenLogsFolder()
 
         public void ToggleFullscreen()
         {
+            if (Environment.TickCount64 - _lastFullscreenToggle < 500)
```
Comment (gdkchan) [inline Ryujinx.Ava/Ui/ViewModels/MainWindowViewModel.cs:933]: In which unit the value 500 is? Ticks? Might be worth adding a constant and maybe a comment here.

File: Ryujinx.Ava/Ui/ViewModels/SettingsViewModel.cs [hunk 7]
```diff
@@ -233,20 +243,31 @@ private unsafe void LoadAvailableGpus()
...
-                    names.Add($"{device.Name} {(device.IsDiscrete ? "(dGpu)" : "")}");
+                    IsVulkanAvailable = false;
+                    GraphicsBackendIndex = 1;
+                }
+                else
+                {
+                    foreach (var device in devices)
+                    {
+                        _gpuIds.Add(device.Id);
+                        names.Add($"{device.Name} {(device.IsDiscrete ? "(dGpu)" : "")}");
```
Comment (gdkchan) [inline Ryujinx.Ava/Ui/ViewModels/SettingsViewModel.cs:257]:
```suggestion
                        names.Add($"{device.Name} {(device.IsDiscrete ? "(dGPU)" : "")}");
```

File: Ryujinx.Ava/Ui/ViewModels/SettingsViewModel.cs [hunk 8]
```diff
@@ -233,20 +243,31 @@ private unsafe void LoadAvailableGpus()
...
                 foreach (var device in VulkanPhysicalDevice.SuitableDevices)
                 {
-                    _gpuIds.Add(VulkanInitialization.StringFromIdPair(device.Value.VendorID, device.Value.DeviceID));
+                    _gpuIds.Add(
+                        VulkanInitialization.StringFromIdPair(device.Value.VendorID, device.Value.DeviceID));
                     var value = device.Value;
                     var name = value.DeviceName;
-                    names.Add($"{Marshal.PtrToStringAnsi((IntPtr)name)} {(device.Value.DeviceType == PhysicalDeviceType.DiscreteGpu ? "(dGpu)" : "")}");
+                    names.Add(
+                        $"{Marshal.PtrToStringAnsi((IntPtr)name)} {(device.Value.DeviceType == PhysicalDeviceType.DiscreteGpu ? "(dGpu)" : "")}");
```
Comment (gdkchan) [inline Ryujinx.Ava/Ui/ViewModels/SettingsViewModel.cs:270]:
```suggestion
                        $"{Marshal.PtrToStringAnsi((IntPtr)name)} {(device.Value.DeviceType == PhysicalDeviceType.DiscreteGpu ? "(dGPU)" : "")}");
```

File: Ryujinx.Graphics.Vulkan/ImageWindow.cs [hunk 9]
```diff
@@ -60,45 +62,57 @@ private void RecreateImages()
...
-                    SharingMode = SharingMode.Exclusive,
-                    InitialLayout = ImageLayout.Undefined,
-                    Flags = ImageCreateFlags.ImageCreateMutableFormatBit
-                };
+                SType = StructureType.ImageCreateInfo,
+                ImageType = ImageType.ImageType2D,
+                Format = Format,
+                Extent =
+                    new Extent3D((uint?)_width,
+                        (uint?)_height, 1),
```
Comment (gdkchan) [inline Ryujinx.Graphics.Vulkan/ImageWindow.cs:96]: That can be moved to the line above.

File: Ryujinx.Graphics.Vulkan/ImageWindow.cs [hunk 10]
```diff
@@ -244,20 +265,42 @@ public override unsafe void Present(ITexture texture, ImageCrop crop, Action<obj
...
 
             _gd.CommandBufferPool.Return(
                 cbs,
                 null,
                 stackalloc[] { PipelineStageFlags.PipelineStageColorAttachmentOutputBit },
                 null);
 
-            var memory = _imageAllocationAuto[_nextImage].GetUnsafe().Memory;
-            var presentInfo = new PresentImageInfo(image.GetUnsafe().Value, memory, _imageSizes[_nextImage], _imageOffsets[_nextImage], _renderFinishedSemaphore, _imageAvailableSemaphore);
+            var j = _nextImage;
```
Comment (gdkchan) [inline Ryujinx.Graphics.Vulkan/ImageWindow.cs:276]: Why is that called `j`?

File: Ryujinx.Graphics.Vulkan/ImageWindow.cs [hunk 11]
```diff
@@ -339,23 +385,89 @@ public override void Dispose()
...
+        private bool _isActive;
+        private readonly Auto<DisposableImage> _image;
+        private readonly Auto<MemoryAllocation> _memory;
+
+        public Image Image
+        {
+            get
+            {
+                _isActive = true;
+                _image.IncrementReferenceCount();
```
Comment (gdkchan) [inline Ryujinx.Graphics.Vulkan/ImageWindow.cs:417]: This seems fragile. Any access of the property would increment the reference count, and the code accessing would have no way to decrement it after. Personally I would do something that enforces the caller to decrement the reference count the same number of times after its done with it, to prevent memory leaks. For example, you can wrap it on a disposable struct and the called can put a `using` to ensure it will be disposed on function or scope exit, and the `Dispose` method can decrement the reference count. I did not look at the caller so it might not be suitable in this case though, but the general idea is that the caller must be aware it has to decrement the reference count.

File: Ryujinx.Graphics.Vulkan/MemoryAllocator.cs [hunk 12]
```diff
@@ -56,7 +56,7 @@ private MemoryAllocation Allocate(int memoryTypeIndex, ulong size, ulong alignme
             return newBl.Allocate(size, alignment, map);
         }
 
-        private static int FindSuitableMemoryTypeIndex(Vk api, PhysicalDevice physicalDevice, uint memoryTypeBits, MemoryPropertyFlags flags)
+        internal static int FindSuitableMemoryTypeIndex(Vk api, PhysicalDevice physicalDevice, uint memoryTypeBits, MemoryPropertyFlags flags)
```
Comment (gdkchan) [inline Ryujinx.Graphics.Vulkan/MemoryAllocator.cs:59]: Why was this changed? On a quick search, I can't find any of the new code calling it.

File: Ryujinx.Graphics.Vulkan/ImageWindow.cs (hunk 11 above)
Comment (emmauss) [reply Ryujinx.Graphics.Vulkan/ImageWindow.cs:412]: The image isn't used and released in one method/scope. Skia holds the image until it flushes, and even then it's not guaranteed that all commands are completed on canvas.Flush. The reference s decremented when the swapchain presents or when the vulkan control is removed from view.

**Comment by gdkchan** [comment]
The bug that I mentioned before still exists. If you play a game that plays an NVDEC video, and start hovering the menus very fast, first you will notice that the video starts lagging a lot. Then eventually, either the game freezes and memory starts leaking, or it just crashes immediately. This only seems to happen when it is playing NVDEC videos (maybe because usually when it does that, it is constantly uploading texture data?).

I tested with validation layers and usually there's nothing out of the ordinary before the crash, but I once got this:
```
00:00:39.306 |E| RenderTimerTickThread Gpu DebugReport: Validation Error: [ UNASSIGNED-Threading-MultipleThreads ] Object 0: handle = 0xfab64d0000000002, type = VK_OBJECT_TYPE_COMMAND_POOL; | MessageID = 0x141cb623 | THREADING ERROR : vkBeginCommandBuffer(): object of type VkCommandPool is simultaneously used in thread 0x4eb4 and thread 0x2848
00:00:39.306 |E| Gpu DebugReport: Validation Error: [ UNASSIGNED-Threading-MultipleThreads ] Object 0: handle = 0xfab64d0000000002, type = VK_OBJECT_TYPE_COMMAND_POOL; | MessageID = 0x141cb623 | THREADING ERROR : vkBeginCommandBuffer(): object of type VkCommandPool is simultaneously used in thread 0x2848 and thread 0x4eb4
```
To reproduce:
Play a game with a long video. I used Shantae and the Seven Sirens. When the video starts playing, click on one of the menus ("File" or whatever) and then keep moving the mouse left and right  on all the menus. Eventually it will crash, usually takes about 10-20 seconds.

**Comment by jduncanator** [comment]
> The bug that I mentioned before still exists. If you play a game that plays an NVDEC video, and start hovering the menus very fast, first you will notice that the video starts lagging a lot. Then eventually, either the game freezes and memory starts leaking, or it just crashes immediately. This only seems to happen when it is playing NVDEC videos (maybe because usually when it does that, it is constantly uploading texture data?).
> 
> 
> 
> I tested with validation layers and usually there's nothing out of the ordinary before the crash, but I once got this:
> 
> ```
> 
> 00:00:39.306 |E| RenderTimerTickThread Gpu DebugReport: Validation Error: [ UNASSIGNED-Threading-MultipleThreads ] Object 0: handle = 0xfab64d0000000002, type = VK_OBJECT_TYPE_COMMAND_POOL; | MessageID = 0x141cb623 | THREADING ERROR : vkBeginCommandBuffer(): object of type VkCommandPool is simultaneously used in thread 0x4eb4 and thread 0x2848
> 
> 00:00:39.306 |E| Gpu DebugReport: Validation Error: [ UNASSIGNED-Threading-MultipleThreads ] Object 0: handle = 0xfab64d0000000002, type = VK_OBJECT_TYPE_COMMAND_POOL; | MessageID = 0x141cb623 | THREADING ERROR : vkBeginCommandBuffer(): object of type VkCommandPool is simultaneously used in thread 0x2848 and thread 0x4eb4
> 
> ```
> 
> To reproduce:
> 
> Play a game with a long video. I used Shantae and the Seven Sirens. When the video starts playing, click on one of the menus ("File" or whatever) and then keep moving the mouse left and right  on all the menus. Eventually it will crash, usually takes about 10-20 seconds.

For me, you don't even need to interact with the menu. If you launch a game that plays a video whilst you're idle on the menu, if you leave it long enough, it will gradually slow down and then just crash. You don't even have to touch the mouse or keyboard.

I can't see anything out of the ordinary except VRAM slowly going up, but the crash doesn't seem to be an OOM - at least from what I can tell.

**Comment by gdkchan** [comment]
As of latest commit, the crash while videos are being played has been fixed here, but now there's a NullReferenceException when closing the window.

File: Ryujinx.Ava/Ui/Backend/Vulkan/VulkanDisplay.cs [hunk 13]
```diff
@@ -15,16 +14,30 @@ internal class VulkanDisplay : IDisposable
...
-        private VulkanDevice _device { get; }
-        private ImageView[] _swapchainImageViews = new ImageView[0];
+        private ImageView[] _swapchainImageViews = Array.Empty<ImageView>();
         private bool _vsyncStateChanged;
         private bool _vsyncEnabled;
+        private bool _surfaceChanged;
+
+        public event EventHandler Presented;
+
+        public bool NotifySurfaceChanged
```
Comment (gdkchan) [inline Ryujinx.Ava/Ui/Backend/Vulkan/VulkanDisplay.cs:31]: This name seems odd for what it does. It is called `NotifySurfaceChanged` but it actually sets `_surfaceChanged` to false. I think this should probably be a regular method, IMO proprties should not do much more than just getting or setting internal state, and I would not expect accessing a property to have "side effects".

File: Ryujinx.Ava/Ui/Controls/RendererControl.cs [hunk 14]
```diff
@@ -13,22 +13,10 @@ internal abstract class RendererControl : Control
...
-        }
-
-        public readonly static StyledProperty<object> ImageProperty =
-            AvaloniaProperty.Register<RendererControl, object>(
-                nameof(Image),
-                0,
-                inherits: true,
-                defaultBindingMode: BindingMode.TwoWay);
-
         protected object Image
```
Comment (gdkchan) [inline Ryujinx.Ava/Ui/Controls/RendererControl.cs:16]: Maybe this could become an auto-property now?

File: Ryujinx.Graphics.Vulkan/ImageWindow.cs [hunk 15]
```diff
@@ -339,23 +365,81 @@ public override void Dispose()
...
-        public ulong MemorySize { get; set; }
-        public ulong MemoryOffset { get; set; }
-        public Semaphore ReadySemaphore { get; }
-        public Semaphore AvailableSemaphore { get; }
-
-        public PresentImageInfo(Image image, DeviceMemory memory, ulong memorySize, ulong memoryOffset, Semaphore readySemaphore, Semaphore availableSemaphore)
+        private readonly Auto<DisposableImage> _image;
+        private readonly Auto<MemoryAllocation> _memory;
+
+        public Image Image
```
Comment (gdkchan) [inline Ryujinx.Graphics.Vulkan/ImageWindow.cs:391]: Those two can be just:
```cs
public Image Image => _image.GetUnsafe().Value;
```

File: Ryujinx.Graphics.Vulkan/WindowBase.cs [hunk 16]
```diff
@@ -10,5 +10,6 @@ internal abstract class WindowBase: IWindow
         public abstract void Dispose();
         public abstract void Present(ITexture texture, ImageCrop crop, Action<object> swapBuffersCallback);
         public abstract void SetSize(int width, int height);
+        public virtual void ChangeVSyncMode(bool vsyncEnabled) { }
```
Comment (gdkchan) [inline Ryujinx.Graphics.Vulkan/WindowBase.cs:13]: Why is that not abstract like the others?

File: Ryujinx/Ui/RendererWidgetBase.cs [hunk 17]
```diff
@@ -402,6 +402,8 @@ public void Render()
                 Device.Gpu.InitializeShaderCache(_gpuCancellationTokenSource.Token);
                 Translator.IsReadyForTranslation.Set();
 
+                Renderer?.Window?.ChangeVSyncMode(Device.EnableDeviceVsync);
```
Comment (gdkchan) [inline Ryujinx/Ui/RendererWidgetBase.cs:405]: Can Window really be null here?

**Review by gdkchan — COMMENTED** [review]
Does this close https://github.com/Ryujinx/Ryujinx/issues/3534 ?

File: Ryujinx/Ui/RendererWidgetBase.cs (hunk 17 above)
Comment (emmauss) [reply Ryujinx/Ui/RendererWidgetBase.cs:405]: No

**Review by gdkchan — APPROVED** [review]
lgtm, thanks.

File: Ryujinx.Ava/Ui/Backend/Vulkan/Surfaces/VulkanSurfaceRenderTarget.cs [hunk 18]
```diff
@@ -1,30 +1,39 @@
...
-            Display = VulkanDisplay.CreateDisplay(platformInterface.Instance, platformInterface.Device,
+            var device = VulkanInitialization.CreateDevice(platformInterface.Api,
+                                                           platformInterface.PhysicalDevice.InternalHandle,
+                                                           platformInterface.PhysicalDevice.QueueFamilyIndex,
+                                                           VulkanInitialization.GetSupportedExtensions(platformInterface.Api, platformInterface.PhysicalDevice.InternalHandle),
+                                                           platformInterface.PhysicalDevice.QueueCount);
+
+            Device = new VulkanDevice(device, platformInterface.PhysicalDevice, platformInterface.Api);
+
+            Display = VulkanDisplay.CreateDisplay(platformInterface.Instance, Device,
```
Comment (AcK77) [inline Ryujinx.Ava/Ui/Backend/Vulkan/Surfaces/VulkanSurfaceRenderTarget.cs:32]: It should be fine on one line

File: Ryujinx.Graphics.Vulkan/ImageWindow.cs [hunk 19]
```diff
@@ -18,87 +20,86 @@ class ImageWindow : WindowBase, IWindow, IDisposable
...
+                SType = StructureType.ImageCreateInfo,
+                ImageType = ImageType.ImageType2D,
+                Format = Format,
+                Extent = new Extent3D((uint?)_width, (uint?)_height, 1),
+                MipLevels = 1,
+                ArrayLayers = 1,
+                Samples = SampleCountFlags.SampleCount1Bit,
+                Tiling = ImageTiling.Optimal,
+                Usage = ImageUsageFlags.ImageUsageColorAttachmentBit | ImageUsageFlags.ImageUsageTransferSrcBit |
+                    ImageUsageFlags.ImageUsageTransferDstBit,
```
Comment (AcK77) [inline Ryujinx.Graphics.Vulkan/ImageWindow.cs:90]: Should be on the above line

File: Ryujinx.Graphics.Vulkan/ImageWindow.cs [hunk 20]
```diff
@@ -205,13 +208,18 @@ public override unsafe void Present(ITexture texture, ImageCrop crop, Action<obj
 
             if (ScreenCaptureRequested)
             {
-                CaptureFrame(view, srcX0, srcY0, srcX1 - srcX0, srcY1 - srcY0, view.Info.Format.IsBgr(), crop.FlipX, crop.FlipY);
+                CaptureFrame(view, srcX0, srcY0, srcX1 - srcX0, srcY1 - srcY0, view.Info.Format.IsBgr(), crop.FlipX,
+                    crop.FlipY);
```
Comment (AcK77) [inline Ryujinx.Graphics.Vulkan/ImageWindow.cs:212]: Should be on the above line

File: Ryujinx.Graphics.Vulkan/ImageWindow.cs [hunk 21]
```diff
@@ -205,13 +208,18 @@ public override unsafe void Present(ITexture texture, ImageCrop crop, Action<obj
...
 
                 ScreenCaptureRequested = false;
             }
 
-            float ratioX = crop.IsStretched ? 1.0f : MathF.Min(1.0f, _height * crop.AspectRatioX / (_width * crop.AspectRatioY));
-            float ratioY = crop.IsStretched ? 1.0f : MathF.Min(1.0f, _width * crop.AspectRatioY / (_height * crop.AspectRatioX));
+            float ratioX = crop.IsStretched
```
Comment (AcK77) [inline Ryujinx.Graphics.Vulkan/ImageWindow.cs:217]: The formatting was better before, on one line (same for ratioY)

**Review by AcK77 — APPROVED** [review]
